README.md
<!---
[![Code Health](https://landscape.io/github/yafp/apparat_launcher/master/landscape.svg?style=flat)](https://landscape.io/github/yafp/apparat_launcher/master)
-->
[![License](https://img.shields.io/badge/license-GPL3-brightgreen.svg)](LICENSE)
[![Issues](https://img.shields.io/github/issues/yafp/apparat_launcher.svg)](https://github.com/yafp/apparat_launcher/issues)
[![Issues Closed](https://img.shields.io/github/issues-closed/yafp/apparat_launcher.svg)](https://github.com/yafp/apparat_launcher/issues?q=is%3Aclosed)
[![PullRequests](https://img.shields.io/github/issues-pr/yafp/apparat_launcher.svg)](https://github.com/yafp/apparat_launcher/pulls)
[![PullRequests-Closed](https://img.shields.io/github/issues-pr-closed/yafp/apparat_launcher.svg)](https://github.com/yafp/apparat_launcher/pulls?q=is%3Apr+is%3Aclosed)
[![Contributors](https://img.shields.io/github/contributors/yafp/apparat_launcher.svg)](https://github.com/yafp/apparat_launcher/graphs/contributors)
[![Build Status](https://travis-ci.org/yafp/apparat_launcher.svg?branch=master)](https://travis-ci.org/yafp/apparat_launcher)


# <a name="top"> apparat_launcher

![logo](https://raw.githubusercontent.com/yafp/apparat_launcher/master/apparat_launcher/gfx/core/128/appIcon.png)


## Important notice
This program is currently in a very early development state.


## <a name="toc">Table of Contents

  * [Description](#description)
  * [Screenshots](#screenshots)
  * [Installation](#installation)
    * [Requirements](#Requirements)
  * [Usage](#usage)
    * [Parameter](#parameter)
    * [Hotkey](#hotkey)
  * [Plugins](#plugins)
    * [Core](#core)
    * [Kill](#kill)
    * [Misc](#misc)
    * [Nautilus](#nautilus)
    * [Password Generator](#passwordgen)
    * [Screenshot](#screenshot)
    * [Search-Internet](#search-internet)
    * [Search-Local](#search-local)
    * [Session](#session)
    * [Shell](#shell)
  * [Changelog](#changelog)
  * [Development](#development)
    * [Contributing](#contributing)
    * [Colors](#colors)
  * [Credits](#credits)


## <a name="description">Description

__apparat_launcher__ is an application launcher for linux. It is developed and tested for Gnome Desktop environments.


## <a name="screenshots">Screenshots
*Launcher UI*

![screenshot](https://raw.githubusercontent.com/yafp/apparat_launcher/master/docs/screenshots_ui/screenshot_ui.png)


*Tray-Icon (Gnome)*

![screenshot](https://raw.githubusercontent.com/yafp/apparat_launcher/master/docs/screenshots_ui/screenshot_trayicon.png)


## <a name="installation">Installation
There is no installation routine so far.

#### <a name="requirements">Requirements
##### Python modules

- ```difflib```
- ```fnmatch```
- ```os```
- ```platform```
- ```psutil```
- ```subprocess```
- ```sys```
- ```webbrowser```
- ```wx```
- ```xdg```
- ```xdg.IconTheme```

##### Linux packages
The following packages are needed:

- ```gnome-screensaver-command```
- ```gnome-session-quit```
- ```systemctl```
- ```xdg-open```
- ```xdotool```


## <a name="usage">Usage
Simply run:
```
./apparat_launcher.py
```


#### <a name="parameter">Parameter
The following command line parameters are available:

| Parameter     | Alias          | Function            |
| ------------- |:---------------| :-------------------|
| ```-d```      | ```--debug```  | Show debug output   |
| ```-h```      | ```--help```   | Show help           |
| ```-v```      | ```--version```| Show version        |

*Debug output*

![screenshot](https://raw.githubusercontent.com/yafp/apparat_launcher/master/docs/screenshots_ui/screenshot_debug_output.png)


#### <a name="hotkey">Hotkey
If you want to use a global/system-wide hotkey to trigger ```apparat_launcher``` (starting, focusing and minimizing) consider using the script ```hotkeyHelperForApparatLauncher.sh```. Simply define a system-wide hotkey which triggers ```hotkeyHelperForApparatLauncher.sh```.

Hotkey definition in [Gnome](https://help.gnome.org/users/gnome-help/stable/keyboard-shortcuts-set.html.en)
* Open the **Activities** overview and start typing Keyboard.
* Click on **Keyboard** to open the panel.
* Select the **Shortcuts** tab.
* Select a category in the left pane, and the row for the desired action (launching ```hotkeyHelperForApparatLauncher.sh```)  on the right. The current shortcut definition will change to **New accelerator…**
* Hold down the desired key combination, or press Backspace to clear.


## <a name="plugins">Plugins
#### <a name="core">Core
Core plugin (can not be disabled)

| Trigger            | Alias              | Parameter      | Function                                        |
|--------------------|:------------------ |:---------------|:------------------------------------------------|
| ```!help```        |                    |                | Open apparat_launcher online documentation      |
| ```!preferences``` | ```!prefs```       |                | Open apparat_launcher preferences               |


#### <a name="kill">Kill
Optional plugin to kill graphical applications using xkill

| Trigger            | Alias              | Parameter      | Function                                        |
|--------------------|:------------------ |:---------------|:------------------------------------------------|
| ```!xkill```       | ```!kill```        |                | Starts xkill                                    |


#### <a name="misc">Misc
Optional plugin

| Trigger            | Alias              | Parameter      | Function                                        |
|--------------------|:------------------ |:---------------|:------------------------------------------------|
| ```!open```        |                    | _path_         | Open file or folder using default application   |


#### <a name="nautilus">Nautilus
Optional plugin which offers quick access to some locations in nautilus

| Trigger          | Alias          | Parameter      | Function                         |
|------------------|:-------------- |:---------------|:---------------------------------|
| ```!goto```      |                | _path_         | Open directory in nautilus       |
| ```!network```   | ```!net```     |                | Show network devices in nautilus |
| ```!recent```    |                |                | Show recent files in nautilus    |
| ```!trash```     |                |                | Show trash in nautilus           |


#### <a name="passwordgen">Password Generator
Optional plugin which offers a simple password generator

| Trigger          | Alias          | Parameter      | Function                         |
|------------------|:-------------- |:---------------|:---------------------------------|
| ```!pw```        | ```!password```|                | A simple password generator      |


#### <a name="screenshot">Screenshot
Optional plugin which offers screenshot functionality

| Trigger          | Alias              | Parameter      | Function                              |
|------------------|:------------------ |:---------------|:--------------------------------------|
| ```!ss```        |                    |                | Selective screenshot (window or area) |
| ```!fs```        |                    |                | Full screenshot                       |


#### <a name="search-internet">Search-Internet
Optional plugin which offers easy access to some popular web-services

| Trigger          | Alias          | Parameter      | Function               |
|------------------|:---------------|:---------------|:---------------------- |
| ```!am```        |                | _searchstring_ | Amazon                 |
| ```!au```        |                | _searchstring_ | Ask Ubuntu             |
| ```!bc```        |                | _searchstring_ | BandCamp               |
| ```!dd```        |                | _searchstring_ | DuckDuckGo             |
| ```!fb```        |                | _searchstring_ | FaceBook               |
| ```!fe```        |                | _searchstring_ | Fefe                   |
| ```!fl```        |                | _searchstring_ | Flickr                 |
| ```!gh```        |                | _searchstring_ | GitHub                 |
| ```!gi```        |                | _searchstring_ | Google Images          |
| ```!gk```        |                | _searchstring_ | Google Keep/Notes      |
| ```!gm```        |                | _searchstring_ | Google Maps            |
| ```!gn```        |                | _searchstring_ | Google News            |
| ```!gs```        |                | _searchstring_ | Google Search          |
| ```!la```        |                | _searchstring_ | LastFM                 |
| ```!re```        |                | _searchstring_ | Reddit                 |
| ```!sc```        |                | _searchstring_ | SoundCloud             |
| ```!se```        |                | _searchstring_ | Stack-Exchange         |
| ```!so```        |                | _searchstring_ | Stack-Overflow         |
| ```!tu```        |                | _searchstring_ | Tumblr                 |
| ```!tw```        |                | _searchstring_ | Twitter                |
| ```!vi```        |                | _searchstring_ | Vimeo                  |
| ```!wi```        |                | _searchstring_ | Wikipedia              |
| ```!yt```        |                | _searchstring_ | YouTube                |


#### <a name="search-local">Search-Local
Optional plugin which offers a file search for $HOME

| Trigger          | Alias          | Parameter      | Function                      |
|------------------|:---------------|:---------------|:----------------------------- |
| ```?```          |                | _searchstring_ | Filesearch for home directory |


#### <a name="session">Session
Optional plugin which offers access to some session commands

| Trigger            | Alias          | Parameter      | Function                    |
|--------------------|:-------------- |:---------------|:----------------------------|
| ```!hibernate```   | ```!sleep```   |                | hibernate the machine       |
| ```!lock```        |                |                | locks the current session   |
| ```!logout```      |                |                | logout from current session |
| ```!reboot```      | ```!restart``` |                | reboot the machine          |
| ```!shutdown```    | ```!halt```    |                | shutdown the machine        |
| ```!screensaver``` | ```!saver```   |                | start screensaver           |


#### <a name="shell">Shell
Optional plugin which allows running terminal commands from within the launcher

| Trigger          | Alias              | Parameter      | Function                            |
|------------------|:------------------ |:---------------|:------------------------------------|
| ```!sh```        |                    | _path_         | Run terminal commands in new window |


## <a name="changelog">Changelog
The file [CHANGELOG.md](CHANGELOG.md) provides an overview of the major changes of this project; for a more detailed look at changes to the code, view individual commits.




## <a name="development">Development

![bomb](https://raw.githubusercontent.com/yafp/apparat_launcher/master/docs/bomb.gif)


### <a name="contributing">Contributing
Please check [CONTRIBUTING.md](.github/CONTRIBUTING.md) for contribution guidelines for this project.


### <a name="colors">Colors
* Gray (#7f8c8d or 127, 140, 141)
* Red (#b92c0c or 185, 44, 12)
* Green (#269c58 or 38, 156, 88)


## <a name="credits">Credits
* Icons [Details](docs/icons.md)
  * [Font Awesome](http://fontawesome.io)
  * [Foundation Icons](http://zurb.com/playground/foundation-icon-fonts-3)
  * [FA2PNG](http://fa2png.io/)


CHANGELOG.md
Changelog
==========

![logo](https://raw.githubusercontent.com/yafp/apparat_launcher/master/apparat_launcher/gfx/core/128/appIcon.png)

# 201706xx
* Tray Menu: Shows application version string now
* UI: All UI items should use the same grey color now


# 20170602
* Plugin Password Generator: changed default button to OK (memorable or not)


# 20170530
* Moved trayicon and traymenu code to new file tray_icon.py
* UI: reduced main window height from 460px to 430px
* UI: Text command has now a highlight for the user input string
* Plugins: Optimized plugin-commands-search regarding when to autocomplete and when not
* Plugin Internet Search: adding support for blog.fefe.de
* tools.py: Added general support for syslog in debug_output                            (Issue: #68)


# 20170523
* Reworking requirements check (is now plugin specific)                                 (Issue: #67)
* Adding first version of apidoc (using epydoc)


# 20170519
* Reset UI function takes now care about icon-size as well
* Project is now using Travis CI for building etc...                                    (Issue: #51)


# 20170518
* Plugins: Added new plugin Kill (using xkill)                                          (Issue: #66)
* Optimized code to detect running application processes


# 20170515
* Plugin-command-search now features only commands from enabled plugins                 (Issue: #62)
* Plugin-command-search is now sorted                                                   (Issue: #63)
* Plugin-command-search got autocomplete now                                            (Issue: #64)
* Plugin-command-search does no autocomplete if user is pressing BackSpace
* Plugin Password Generator: UI is now always resetted after finishing the function
* Debug Output: Now shows the source (.py) file as well, not only function name)        (Issue: #65)


# 20170511
* Plugins: Fill searchresults with matching plugin-triggers (if input starts with !)    (Issue: #59)
* Plugin Password Generator: Add support for memorable passwords                        (Issue: #60)
* Plugin Password Generator: Now generates 5 passwords on execution to offer a choice   (Issue: #61)
* Renamed apparatHotkeyHelper.sh to hotkeyHelperForApparatLauncher


# 20170509
* Plugin Internet Search: Add support for Google Keep/Notes (!gk)                       (Issue: #53)
* Plugin Password Generator: Added initial version of new plugin                        (Issue: #54)
* Plugins: Fix wrong result count while incomplete plugin command                       (Issue: #55)
* Command button is now displaying appIcon if disabled                                  (Issue: #56)
* Parameter button is now showing search-icon if disabled                               (Issue: #57)
* Parameter button now shows execute icon if search results > 1 (formerly blank)        (ISsue: #58)


# 20170508
* Fix left-over from #50 in apparat.py (get_icon) and plugin_nautilus.py
* Finishing rename to apparat_launcher.py                                               (Issue: #42)
* UI: Transparency change now via preferences (before: config.py)                       (Issue: #52)


# 20170506
* Minor fix for apparatHotkeyHelper.sh
* Minor changes based on pylint
* Plugin Shell:
    Moved from xterm to x-terminal-emulator as default
    Opens now a new default terminal and executes the related command
* Optimized method get_icon in apparat-launcher.py
* Readme.md: Fixing landscape link                                                      (Issue: #42)
* Ini: Adjusted ini folder and ini file to new project name                             (Issue: #42)
* Project: renaming once again from apparat-launcher to apparat_launcher                (Issue: #42)
    (apparat-launcher conflicts with naming conventions)
* UI: Status icon is hidden by default. Only displayed if there is an OK or ERROR state
* UI: Icon size change now via preferences (before: config.py)                          (Issue: #50)


# 20170427
* Add requirements check to apparatHotkeyHelper.sh
* Plugin Internet Search: Move from single to dual-letter triggers                      (Issue: #40)
* Plugin Internet Search: Add support for Ask Ubuntu via !au                            (Issue: #41)
* Plugin Internet Search: Add support for GitHub via !gh                                (Issue: #39)
* Plugin Internet Search: Add support for Facebook via !fb                              (Issue: #43)
* Renamed project from apparat to apparat-launcher                                      (Issue: #42)
* Plugin Internet Search: Add support for Flickr via !fl                                (Issue: #44)
* Plugin Internet Search: Add support for Google Images via !gi                         (Issue: #45)
* Plugin Internet Search: Add support for Google News via !gn                           (Issue: #46)
* Plugin Internet Search: Add support for DuckDuckGo Search via !dd                     (Issue: #47)
* Plugin Internet Search: Add support for tumblr via !tu                                (Issue: #48)
* Preferences: Add description text for each plugin on plugin-preference-tab1           (Issue: #49)


# 20170415
* Prefs: Plugins can now be enabled or disabled.                                        (Issue: #37)
* Plugins: Create new plugin core which handles !prefs, !preferences and !helper        (Issue: #38)


# 20170413
* Plugin Misc: Added ~ support for !open command                                        (Issue: #35)
* Plugin Nautilus: Added ~ support for !goto command                                    (Issue: #36)
* Improved check for required linux packages
* Improved get_icon method in case of .svg results


# 20170412
* Preferences: Checkboxes on plugin-tab are not longer disabled,
    but state can't be changed. All plugins are always enabled.                         (Issue: #33)
* Ini: Added basic validation (existing sections and options)                           (Issue: #13)

apparat/prefs.py
#!/usr/bin/python
"""apparat - an application launcher for linux"""

# -----------------------------------------------------------------------------------------------
# IMPORTS
# -----------------------------------------------------------------------------------------------

## built-in modules
import wx

## apparat
import constants
import ini
import tools


# -----------------------------------------------------------------------------------------------
# PREFERENCE WINDOW
# -----------------------------------------------------------------------------------------------
class PreferenceWindow(wx.Frame):

    """Class for Preference Window"""

    def __init__(self, parent, idd):
        """Initialize the preference window"""
        ## define style of preference window
        pref_window_style = (wx.SYSTEM_MENU | wx.CAPTION | wx.CLOSE_BOX | wx.CLIP_CHILDREN | wx.STAY_ON_TOP | wx.FRAME_NO_TASKBAR)
        wx.Frame.__init__(self, parent, idd, constants.APP_NAME+' Preferences', size=(600, 700), style=pref_window_style)

        ## Create a panel and notebook (tabs holder)
        p = wx.Panel(self)
        nb = wx.Notebook(p)

        ## Create the tab windows
        tab1 = UITabGeneral(nb)
        tab2 = UITabStatistics(nb)
        tab3 = UITabPluginCommands(nb)

        ## Add the windows to tabs and name them.
        nb.AddPage(tab1, "General ")
        nb.AddPage(tab2, "Statistics ")
        nb.AddPage(tab3, "Plugins ")

        ## Set noteboook in a sizer to create the layout
        sizer = wx.BoxSizer()
        sizer.Add(nb, 1, wx.EXPAND)
        p.SetSizer(sizer)

        wx.Frame.CenterOnScreen(self) # center the pref window

        self.Bind(wx.EVT_CLOSE, self.close_preference_ui)


    def close_preference_ui(self, event):
        """Closes the preference window"""
        tools.debug_output('close_preference_ui', 'starting', 1)
        tools.debug_output('close_preference_ui', 'Event: '+str(event), 1)
        self.Destroy() # close the pref UI


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# PREFERENCE-TABS- TABS - # https://pythonspot.com/wxpython-tabs/
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
class UITabGeneral(wx.Panel):

    """Preference Window - Tab: General"""

    def __init__(self, parent):
        """Inits the general tab"""
        wx.Panel.__init__(self, parent)

        ## show language
        cur_ini_value_for_language = ini.read_single_value('Language', 'lang')          # get current value from ini
        t = wx.StaticText(self, -1, "Language: ", (20, 20))

        languages = [cur_ini_value_for_language]
        combo_box_style = wx.CB_READONLY
        ui__cb_languages = wx.ComboBox(self, wx.ID_ANY, u'', wx.DefaultPosition, wx.Size(100, 30), languages, style=combo_box_style)
        ui__cb_languages.SetFont(wx.Font(12, wx.MODERN, wx.NORMAL, wx.NORMAL, False, u'Sans'))
        ui__cb_languages.SetValue(languages[0])

        ## Hide UI
        self.cb_enable_hide_ui = wx.CheckBox(self, -1, 'Hide UI after command execution ', (20, 60))
        cur_ini_value_for_hide_ui_after_command_execution = ini.read_single_value('General', 'hide_ui_after_command_execution')          # get current value from ini
        if cur_ini_value_for_hide_ui_after_command_execution == "False":
            self.cb_enable_hide_ui.SetValue(False)
        else:
            self.cb_enable_hide_ui.SetValue(True)
        wx.EVT_CHECKBOX(self, self.cb_enable_hide_ui.GetId(), self.prefs_general_toggle_hide_ui)

        ## Layout
        general_sizer = wx.BoxSizer(wx.VERTICAL) # define layout container
        general_sizer.AddSpacer(10)
        general_sizer.Add(t, 0, wx.ALL, border=10)
        general_sizer.Add(ui__cb_languages, 0, wx.ALL, border=10)
        general_sizer.AddSpacer(10)
        general_sizer.Add(self.cb_enable_hide_ui, 0, wx.ALL, border=10)
        self.SetSizer(general_sizer)


    def prefs_general_toggle_hide_ui(self, event):
        """Toggle the general pref: hide_ui"""
        tools.debug_output('prefs_general_toggle_hide_ui', 'Preference - General - Hide UI: '+str(event), 1)
        if self.cb_enable_hide_ui.GetValue() is True:
            tools.debug_output('prefs_general_toggle_hide_ui', 'Enabled', 1)
            ini.write_single_value('General', 'hide_ui_after_command_execution', "True") # update preference value
        else:
            tools.debug_output('prefs_general_toggle_hide_ui', 'Disabled', 1)
            ini.write_single_value('General', 'hide_ui_after_command_execution', "False") # update preference value



class UITabStatistics(wx.Panel):

    """Preference Window - Tab: Statistics - Shows usage stats"""

    def __init__(self, parent):
        """Inits the statistics tab"""
        wx.Panel.__init__(self, parent)

        ## show app start counter
        cur_ini_value_for_apparat_started = ini.read_single_value('Statistics', 'apparat_started')          # get current value from ini
        t1 = wx.StaticText(self, -1, "Apparat started:\t\t\t"+cur_ini_value_for_apparat_started, (20, 20))

        ## show execute counter
        cur_ini_value_for_command_executed = ini.read_single_value('Statistics', 'command_executed')          # get current value from ini
        t2 = wx.StaticText(self, -1, "Command executed:\t\t"+cur_ini_value_for_command_executed, (20, 40))

        ## show plugin trigger count
        cur_ini_value_for_plugin_executed = ini.read_single_value('Statistics', 'plugin_executed')          # get current value from ini
        t3 = wx.StaticText(self, -1, "Plugins executed:\t\t\t"+cur_ini_value_for_plugin_executed, (20, 60))

        statistics_sizer = wx.BoxSizer(wx.VERTICAL) # define layout container
        statistics_sizer.AddSpacer(10)
        statistics_sizer.Add(t1, 0, wx.ALL, border=10)
        statistics_sizer.AddSpacer(10)
        statistics_sizer.Add(t2, 0, wx.ALL, border=10)
        statistics_sizer.AddSpacer(10)
        statistics_sizer.Add(t3, 0, wx.ALL, border=10)
        self.SetSizer(statistics_sizer)



class UITabPluginCommands(wx.Panel):

    """Preference Window - Tab: Commands- Shows available plugins"""

    def __init__(self, parent): # pylint:disable=too-many-statements
        """Inits the plugin-commands tab"""
        wx.Panel.__init__(self, parent)

        ## Plugin: core (can not be disabled)
        self.cb_enable_plugin_core = wx.CheckBox(self, -1, 'Core', (20, 60))
        self.cb_enable_plugin_core.SetLabel('core')
        self.cb_enable_plugin_core.SetToolTipString(u'Offers !help !prefs and !preferences.Plugin can not be disabled.')
        self.cb_enable_plugin_core.SetValue(True)
        self.cb_enable_plugin_core.Bind(wx.EVT_CHECKBOX, self.on_plugin_checkbox_click) # changing the checkbox change

        ## Plugin: Local search
        self.cb_enable_plugin_local_search = wx.CheckBox(self, -1, 'Local-Search', (20, 60))
        self.cb_enable_plugin_local_search.SetLabel('search_local')
        self.cb_enable_plugin_local_search.SetToolTipString(u'Enables search for files and folders in users home directory')
        cur_ini_value_for_plugin_search_local = ini.read_single_value('Plugins', 'enable_plugin_search_local')          # get current value from ini
        if cur_ini_value_for_plugin_search_local == 'True':
            self.cb_enable_plugin_local_search.SetValue(True)
        else:
            self.cb_enable_plugin_local_search.SetValue(False)
        self.cb_enable_plugin_local_search.Bind(wx.EVT_CHECKBOX, self.on_plugin_checkbox_change) # changing the checkbox change

        ## Plugin: Internet search
        self.cb_enable_plugin_internet_search = wx.CheckBox(self, -1, 'Internet-Search', (20, 60))
        self.cb_enable_plugin_internet_search.SetLabel('search_internet')
        self.cb_enable_plugin_internet_search.SetToolTipString(u'Enables search for several popular web-services')
        cur_ini_value_for_plugin_search_internet = ini.read_single_value('Plugins', 'enable_plugin_search_internet')          # get current value from ini
        if cur_ini_value_for_plugin_search_internet == 'True':
            self.cb_enable_plugin_internet_search.SetValue(True)
        else:
            self.cb_enable_plugin_internet_search.SetValue(False)
        self.cb_enable_plugin_internet_search.Bind(wx.EVT_CHECKBOX, self.on_plugin_checkbox_change) # changing the checkbox change

        ## Plugin: Nautilus
        self.cb_enable_plugin_nautilus = wx.CheckBox(self, -1, 'Nautilus', (20, 60))
        self.cb_enable_plugin_nautilus.SetLabel('nautilus')
        self.cb_enable_plugin_nautilus.SetToolTipString(u'Enables quick access to some nautilus locations/places')
        cur_ini_value_for_plugin_nautilus = ini.read_single_value('Plugins', 'enable_plugin_nautilus')          # get current value from ini
        if cur_ini_value_for_plugin_nautilus == 'True':
            self.cb_enable_plugin_nautilus.SetValue(True)
        else:
            self.cb_enable_plugin_nautilus.SetValue(False)
        self.cb_enable_plugin_nautilus.Bind(wx.EVT_CHECKBOX, self.on_plugin_checkbox_change) # changing the checkbox change

        ## Plugin: Screenshot
        self.cb_enable_plugin_screenshot = wx.CheckBox(self, -1, 'Screenshot', (20, 60))
        self.cb_enable_plugin_screenshot.SetLabel('screenshot')
        self.cb_enable_plugin_screenshot.SetToolTipString(u'Enables simple screenshot functions')
        cur_ini_value_for_plugin_screenshot = ini.read_single_value('Plugins', 'enable_plugin_screenshot')          # get current value from ini
        if cur_ini_value_for_plugin_screenshot == 'True':
            self.cb_enable_plugin_screenshot.SetValue(True)
        else:
            self.cb_enable_plugin_screenshot.SetValue(False)
        self.cb_enable_plugin_screenshot.Bind(wx.EVT_CHECKBOX, self.on_plugin_checkbox_change) # changing the checkbox change

        ## Plugin: Session
        self.cb_enable_plugin_session = wx.CheckBox(self, -1, 'Session', (20, 60))
        self.cb_enable_plugin_session.SetLabel('session')
        self.cb_enable_plugin_session.SetToolTipString(u'Enables several session commands')
        cur_ini_value_for_plugin_session = ini.read_single_value('Plugins', 'enable_plugin_session')          # get current value from ini
        if cur_ini_value_for_plugin_session == 'True':
            self.cb_enable_plugin_session.SetValue(True)
        else:
            self.cb_enable_plugin_session.SetValue(False)
        self.cb_enable_plugin_session.Bind(wx.EVT_CHECKBOX, self.on_plugin_checkbox_change) # changing the checkbox change

        ## Plugin: Shell
        self.cb_enable_plugin_shell = wx.CheckBox(self, -1, 'Shell', (20, 60))
        self.cb_enable_plugin_shell.SetLabel('shell')
        self.cb_enable_plugin_shell.SetToolTipString(u'Enable executing shell commands')
        cur_ini_value_for_plugin_shell = ini.read_single_value('Plugins', 'enable_plugin_shell')          # get current value from ini
        if cur_ini_value_for_plugin_shell == 'True':
            self.cb_enable_plugin_shell.SetValue(True)
        else:
            self.cb_enable_plugin_shell.SetValue(False)
        self.cb_enable_plugin_shell.Bind(wx.EVT_CHECKBOX, self.on_plugin_checkbox_change) # changing the checkbox change

        ## Plugin: Misc
        self.cb_enable_plugin_misc = wx.CheckBox(self, -1, 'Misc', (20, 60))
        self.cb_enable_plugin_misc.SetToolTipString(u'Enable other stuff')
        self.cb_enable_plugin_misc.SetLabel('misc')
        cur_ini_value_for_plugin_misc = ini.read_single_value('Plugins', 'enable_plugin_misc')          # get current value from ini
        if cur_ini_value_for_plugin_misc == 'True':
            self.cb_enable_plugin_misc.SetValue(True)
        else:
            self.cb_enable_plugin_misc.SetValue(False)
        self.cb_enable_plugin_misc.Bind(wx.EVT_CHECKBOX, self.on_plugin_checkbox_change) # changing the checkbox change

        ## Link to plugin commands description
        wxHyperlinkCtrl = wx.HyperlinkCtrl(self, -1, 'Plugin command details', constants.APP_URL+'#plugins')



        ## -----------------------
        ## Layout
        ## -----------------------
        pref_sizer = wx.BoxSizer(wx.VERTICAL) # define layout container
        pref_sizer.AddSpacer(10)

        ## core
        pref_sizer.Add(self.cb_enable_plugin_core, 0, wx.ALL, border=10)
        pref_sizer.AddSpacer(5)

        ## misc
        pref_sizer.Add(self.cb_enable_plugin_misc, 0, wx.ALL, border=10)
        pref_sizer.AddSpacer(5)

        ## Nautilus
        pref_sizer.Add(self.cb_enable_plugin_nautilus, 0, wx.ALL, border=10)
        pref_sizer.AddSpacer(5)

        ## screenshot
        pref_sizer.Add(self.cb_enable_plugin_screenshot, 0, wx.ALL, border=10)
        pref_sizer.AddSpacer(5)

        ## search internet
        pref_sizer.Add(self.cb_enable_plugin_internet_search, 0, wx.ALL, border=10)
        pref_sizer.AddSpacer(5)

        ## search local
        pref_sizer.Add(self.cb_enable_plugin_local_search, 0, wx.ALL, border=10)
        pref_sizer.AddSpacer(5)

        ## Session
        pref_sizer.Add(self.cb_enable_plugin_session, 0, wx.ALL, border=10)
        pref_sizer.AddSpacer(5)

        ## Shell
        pref_sizer.Add(self.cb_enable_plugin_shell, 0, wx.ALL, border=10)
        pref_sizer.AddSpacer(20)

        ## Hyperlink to docs
        pref_sizer.Add(wxHyperlinkCtrl, 0, wx.ALL, border=10)

        self.SetSizer(pref_sizer)


    def on_plugin_checkbox_click(self, evt):
        """Handle plugin checkbox click - prevent changing values"""
        e_obj = evt.GetEventObject()
        e_obj.SetValue(not e_obj.GetValue()) # plugin checkboxes are always TRUE/checked
        wx.MessageBox('Plugin can not be disabled', 'Error', wx.OK | wx.ICON_WARNING)


    def on_plugin_checkbox_change(self, event):
        """Triggered if a checkbox state is changing and writes the new state to ini"""
        btn = event.GetEventObject().GetLabel()
        value = event.GetEventObject().GetValue()

        if value is True:
            tools.debug_output('on_plugin_checkbox_change', 'Enabled Plugin: '+btn, 1)
            ini.write_single_value('Plugins', 'enable_plugin_'+btn, 'True')
        else:
            tools.debug_output('on_plugin_checkbox_change', 'Disabled Plugin: '+btn, 1)
            ini.write_single_value('Plugins', 'enable_plugin_'+btn, 'False')
